ath79: add support for MikroTik RB951Ui-2HnD

kempniu opened the pull request:
Mikrotik RB951Ui-2HnD is a wireless SOHO router that was previously supported by the ar71xx target, see commit d19b868b12 ("ar71xx: Add support for MikroTik RB951Ui-2HnD").

Specifications
--------------

  - SoC: Atheros AR9344 (600 MHz)
  - RAM: 128 MB (2x 64 MB)
  - Storage: 128 MB NAND flash (various manufacturers)
  - Ethernet: Atheros AR8229 switch, 5x 10/100 Mbit/s
      - 1x PoE in (port 1, 8-30 V input)
      - 1x PoE out (port 5, 500 mA output)
  - Wireless: Atheros AR9340 (802.11b/g/n)
  - USB: 2.0 (1A)
  - 9x LED:
      - 1x power (green, not configurable)
      - 1x user (green)
      - 5x FE ports (green)
      - 1x wireless (green)
      - 1x PoE out (red)
  - 1x button (restart)

See https://mikrotik.com/product/RB951Ui-2HnD for more details.

Flashing
--------

TFTP boot initramfs image and then perform sysupgrade.  Follow common MikroTik procedures at https://openwrt.org/toh/mikrotik/common.

~~Notes~~
-----

~~The DTS file does not inherit from `target/linux/ath79/dts/ar9344_mikrotik_routerboard.dtsi` because the LEDs listed in that file use the same GPIO numbers as those in this PR, but have names that would have been confusing - RB951Ui-2HnD does not have any RSSI LEDs, only Ethernet port status LEDs.~~ (Addressed as of 102c16b79e440cd5927750f9de10bdf7456b0dc1.)

john-tho commented:
> The DTS file does not inherit from target/linux/ath79/dts/ar9344_mikrotik_routerboard.dtsi because the LEDs

Note first, I am not a maintainer, so don't spend too much time changing things for my opinions.
If there are minimal differences between the DTS, either cut the bits you don't use, and put them in an intermediate dtsi for the existing devices, or use `/delete-node/` & `/delete-property/` for your new DTS.

Using fixed nvmem cells for the MAC address _should_ work provided Mikrotik stuck to the pattern of `Hard`, `tag@0x4: ID:0x1a len:0x04`, `tag@0xc: ID:0x04 len:0x08` (MAC address). It follows this pattern on all the devices I have seen, but does not need to because `hard_config` is a type-length-value store. Better we build an nvmem-layouts parser for it (very similar onie TLV parser is _nearly_ in kernel: https://lore.kernel.org/all/20230308153200.682248-20-miquel.raynal@bootlin.com/, adapted with the logic from https://github.com/openwrt/openwrt/blob/master/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c). Otherwise, safer to do what the existing Mikrotik devices do, and get the MAC address from the rb_sysfs platform driver.

Cheers

kempniu commented:
> Note first, I am not a maintainer, so don't spend too much time changing things for my opinions. If there are minimal differences between the DTS, either cut the bits you don't use, and put them in an intermediate dtsi for the existing devices, or use `/delete-node/` & `/delete-property/` for your new DTS.

Ah, neat, thanks, I was unaware of these!  In 102c16b79e440cd5927750f9de10bdf7456b0dc1, I used `/delete-node/` in two places to enable inheriting from `target/linux/ath79/dts/ar9344_mikrotik_routerboard.dtsi` - the resulting DTS file is about 20% smaller than the original one, so I guess it makes sense to keep things this way.  Thanks for suggesting this, @john-tho.

> Using fixed nvmem cells for the MAC address _should_ work provided Mikrotik stuck to the pattern of `Hard`, `tag@0x4: ID:0x1a len:0x04`, `tag@0xc: ID:0x04 len:0x08` (MAC address). It follows this pattern on all the devices I have seen, but does not need to because `hard_config` is a type-length-value store. Better we build an nvmem-layouts parser for it (very similar onie TLV parser is _nearly_ in kernel: https://lore.kernel.org/all/20230308153200.682248-20-miquel.raynal@bootlin.com/, adapted with the logic from https://github.com/openwrt/openwrt/blob/master/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c). Otherwise, safer to do what the existing Mikrotik devices do, and get the MAC address from the rb_sysfs platform driver.

I only used `nvmem-cells` to prevent these warnings from appearing in the kernel ring buffer at boot:

    ag71xx 19000000.eth: invalid MAC address, using random address

But the updates to `target/linux/ath79/mikrotik/base-files/etc/board.d/02_network` present in this PR will indeed cause the base MAC address stored on the board to be extracted from sysfs and applied via UCI, so if that is acceptable, it sure works for me; I dropped everything related to `nvmem-cells` in 102c16b79e440cd5927750f9de10bdf7456b0dc1 to keep things simple.

102c16b79e440cd5927750f9de10bdf7456b0dc1 also addresses a minor problem stemming from diacritics in my name that was spotted by [one of the branch status checks](https://github.com/openwrt/openwrt/actions/runs/4472410563/jobs/7860098273).

f00b4r0 commented:
> Using fixed nvmem cells for the MAC address _should_ work provided Mikrotik stuck to the pattern of `Hard`, `tag@0x4: ID:0x1a len:0x04`, `tag@0xc: ID:0x04 len:0x08` (MAC address). It follows this pattern on all the devices I have seen, but does not need to because `hard_config` is a type-length-value store. Better we build an nvmem-layouts parser for it (very similar onie TLV parser is _nearly_ in kernel: https://lore.kernel.org/all/20230308153200.682248-20-miquel.raynal@bootlin.com/, adapted with the logic from https://github.com/openwrt/openwrt/blob/master/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c). 

Shameless plug: I contemplated doing that and then eventually renounced, because it won't work in all cases. Specific example: the hap-ac2 comes in two known hard_config variants, one with both wlan calibration blobs in the same tag with an offset, and one where each blob gets its own tag, resulting in the [current mechanism](https://github.com/openwrt/openwrt/blob/3ab670b24eee8adef77a71d9d11ee362c1668b26/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata#L118-L121) where we try both cases. I don't think NVMEM can be made to work transparently with that, meaning we would still need to resort to userspace parsing as we do now, and thus it feels like the effort of moving the rb platform driver to NVMEM isn't warranted. But you're welcome to try and convince me otherwise :)

f00b4r0 commented on target/linux/ath79/image/mikrotik.mk:
```diff
@@ -48,6 +48,16 @@ define Device/mikrotik_routerboard-922uags-5hpacd
+  DEVICE_MODEL := RouterBOARD 951Ui-2HnD
+  DEVICE_PACKAGES += kmod-usb-ohci kmod-usb2
+  SUPPORTED_DEVICES += rb-951ui-2hnd
+  IMAGE/sysupgrade.bin = append-kernel | sysupgrade-tar | append-metadata
```
Very minor nit: I would instead add a `mikrotik_nandyafut` (or somesuch) common recipe in `common-mikrotik.mk`, and use that as `$(Device/mikrotik_nandyafut)` instead of `$(Device/mikrotik_nand)`.

This will avoid cluttering every device that is moved to this system and will make it easy to spot the ones that have been switched over and/or mass switch the remaining ones over.

kempniu replied:
Ah, yes, this is definitely a good idea, thanks.  I guess the `IMAGE/sysupgrade.bin` line you commented on is simply a leftover from my initial attempts to get things working that I never touched again because I did not need to.  But the approach you suggested is clearly superior in the general case.

While you pointed this out in this PR, it seems to me that adding a new `Device/...` definition to `target/linux/ath79/image/common-mikrotik.mk` should rather happen in #12225 and this PR should only use that definition.  I will assume that approach for the time being, but feel free to let me know if you disagree.

kempniu replied:
I added the device template suggested above in dacf09232647d2742fedc2f59f96d93e7e7732c2, which is part of #12225.  ee0b5862e30b55f03e7f76d47e1a83ae50c6f719, i.e. the current version of this PR, uses that template.

I am going to resolve this discussion, but obviously please let me know if you have any further comments.  Thanks!

kempniu replied:
This was approached in a different after all, see [elsewhere](https://github.com/openwrt/openwrt/pull/12225#discussion_r1154366821).

Leo-PL commented:
@kempniu one general remark: is pinning to the -2HnD variant required? Most of 9xx Routerboards have 2GHz and 5Ghz variants which from software perspective, differ only by radio calibration data. I would drop the -2HnD suffix, because it may be misleading, or add an -5HnD suffix, to explicitly denote that both variants are supported.

The same is the case, for example with RB912UAG and RB911G. For example - a RB911g-5HPnD works fine on an RB912UAG-2HPnD image for me.

f00b4r0 commented:
> @kempniu one general remark: is pinning to the -2HnD variant required? Most of 9xx Routerboards have 2GHz and 5Ghz variants which from software perspective, differ only by radio calibration data. I would drop the -2HnD suffix, because it may be misleading, or add an -5HnD as an alternate suffix, using DEVICE_ALTn_VENDOR and DEVICE_ALTn_MODEL attributes in image makefile, to explicitly denote that both variants are supported.
> 
> The same is the case, for example with RB912UAG and RB911G. For example - a RB911g-5HPnD works fine on an RB912UAG-2HPnD image for me.

This is the case for many of these devices and so far the answer seems to have been to use a common DTSI and per-device DTS that essentially only set the model name. See for instance how it's done with the various TP-Link TL-WRXXX.

My understanding of the DEVICE_ALTn_* is that they're supposed to be used only for slightly different marketing names of the exact same device, which wouldn't align with the situation here.

Finally dropping the suffix is a very bad idea because there is for example a RB951Ui-2nD which is a NOR device (that we already support). So please don't do this.

Leo-PL commented:
> > @kempniu one general remark: is pinning to the -2HnD variant required? Most of 9xx Routerboards have 2GHz and 5Ghz variants which from software perspective, differ only by radio calibration data. I would drop the -2HnD suffix, because it may be misleading, or add an -5HnD as an alternate suffix, using DEVICE_ALTn_VENDOR and DEVICE_ALTn_MODEL attributes in image makefile, to explicitly denote that both variants are supported.
> > The same is the case, for example with RB912UAG and RB911G. For example - a RB911g-5HPnD works fine on an RB912UAG-2HPnD image for me.
> 
> This is the case for many of these devices and so far the answer seems to have been to use a common DTSI and per-device DTS that essentially only set the model name. See for instance how it's done with the various TP-Link TL-WRXXX.
> 
> My understanding of the DEVICE_ALTn_* is that they're supposed to be used only for slightly different marketing names of the exact same device, which wouldn't align with the situation here.

Not exactly, @hauke suggested to me to use it for differentiating between Ruckus ZF7343 and ZF7363, which can operate on the same image, only difference being single vs dual radios on PCI bus. In the earlier case the absent radio simply won't probe. ZF7341 would get a different device tree only to disable absent Ethernet ports, which couldn't be detected otherwise.

> 
> Finally dropping the suffix is a very bad idea because there is for example a RB951Ui-2nD which is a NOR device (that we already support). So please don't do this.

I agree, distinction between -2HnD and -2nD needs to be clear. We have a few options:
- replace 2/5 with x (would make harder to search for the device),
- use DEVICE_ALTn syntax and have common DT compatible string to reduce code duplication, having [2/5] in the machine name in the device tree,
- use the marketing name if there is one instead of model specification, like for hAP or 911 lite - not every device has that,
- create single .dtsi and .dts with different names - will effectively duplicate the images and resources on buildbots, unless there are any userspace distinctions requiring to use different compatible string set in the single .dts.

I think I've seen every option here in the past. To me using DEVICE_ALTn syntax is the most clean now, but it's just an opinion. Still I believe, that a clear statement is needed, that both band variants are supported, unless they differ enough to require separate images, because software needs to behave differently on them.

I'd gladly hear @hauke's take on that.

f00b4r0 commented:
> > My understanding of the DEVICE_ALTn_* is that they're supposed to be used only for slightly different marketing names of the exact same device, which wouldn't align with the situation here.
> 
> Not exactly, @hauke suggested to me to use it for differentiating between Ruckus ZF7343 and ZF7363, which can operate on the same image, only difference being single vs dual radios on PCI bus. In the earlier case the absent radio simply won't probe. ZF7341 would get a different device tree only to disable absent Ethernet ports, which couldn't be detected otherwise.

Seems like there's no clear policy on DEVICE_ALT usage, it would be good to clarify.
Also IIRC (could be wrong) there's only 1 or 2 DEVICE_ALT* allowable yet there are some Mikrotik device that would require more (e.g. the [LHG-HB platform](https://openwrt.org/toh/mikrotik/lhg-hb_platform)).

> > Finally dropping the suffix is a very bad idea because there is for example a RB951Ui-2nD which is a NOR device (that we already support). So please don't do this.
> 
> I agree, distinction between -2HnD and -2nD needs to be clear. We have a few options:
> 
> * replace 2/5 with x (would make harder to search for the device),

No, please don't do that. The -2nD vs -2HnD is already a very subtle difference that not all users would catch on. Adding placeholder characters would only make this worse. Mikrotik naming is already a nightmare, and they even have a few devices with the same model # yet different innards. Let's not confuse users even more.

> * use DEVICE_ALTn syntax and have common DT compatible string to reduce code duplication, having [2/5] in the machine name in the device tree,

See above limitation, which is possibly the reason why TL-WRXXX all have a separate image that only differ in model name.

> * use the marketing name if there is one instead of model specification, like for hAP or 911 lite - not every device has that,
> * create single .dtsi and .dts with different names - will effectively duplicate the images and resources on buildbots, unless there are any userspace distinctions requiring to use different compatible string set in the single .dts.

That's (unfortunately) what seems to be the current policy. The image duplication was essentially inevitable as soon as we switched to DTS in ath79.

> I think I've seen every option here in the past. To me using DEVICE_ALTn syntax is the most clean now, but it's just an opinion. Still I believe, that a clear statement is needed, that both band variants are supported, unless they differ enough to require separate images, because software needs to behave differently on them.

Good luck with that. Take a look at #3270: this has been discussed ad nauseam and no clear conclusion was ever reached. So I do not think it's correct to hold this particular PR just because of this "issue".

Case in point, I personally run the lhg-2nd image on my sxtsq-2nd, and have [documented that it works](https://openwrt.org/toh/mikrotik/lhg-hb_platform#supported_versions) :P

kempniu commented:
@Leo-PL, thanks for bringing this up as there is one more data point here that should be taken into account.

While I am unaware of any 5GHz variant of the [RB951Ui-2HnD](https://mikrotik.com/product/RB951Ui-2HnD), it does have a "twin" board called [RB951G-2HnD](https://mikrotik.com/product/RB951G-2HnD) that has Gigabit Ethernet ports and was also previously supported by the `ar71xx` target.  I own both of these and I plan to also add support for the RB951G-2HnD in a separate PR.  I have not yet had the time to look particularly closely at that other model, but from what I can tell right now, it *boots* using the DTS proposed for the RB951Ui-2HnD in this PR, but networking does not *work* on it with that DTS due to a different switch configuration.

Does the above influence the recommended course of action wrt naming DTS files?  It might take some time before I can sit down to the RB951G-2HnD to make sure all of its components are operational with a DTS like the one proposed in this PR, but it cannot hurt to plan ahead on the assumption that it will eventually work with something very similar to what was proposed here.

f00b4r0 commented:
> @Leo-PL, thanks for bringing this up as there is one more data point here that should be taken into account.
> 
> While I am unaware of any 5GHz variant of the [RB951Ui-2HnD](https://mikrotik.com/product/RB951Ui-2HnD),

I don't think it exists either.

> it does have a "twin" board called [RB951G-2HnD](https://mikrotik.com/product/RB951G-2HnD) that has Gigabit Ethernet ports and was also previously supported by the `ar71xx` target. I own both of these and I plan to also add support for the RB951G-2HnD in a separate PR. I have not yet had the time to look particularly closely at that other model, but from what I can tell right now, it _boots_ using the DTS proposed for the RB951Ui-2HnD in this PR, but networking does not _work_ on it with that DTS due to a different switch configuration.
> 
> Does the above influence the recommended course of action wrt naming DTS files? 

If there are enough similarities, create a DTSI with the common bits and include it from the two separate device DTS that will adjust the differences. There are existing examples for Mikrotik (and other) ath79 devices in the repo.

There is no strong policy for DTSI naming, it's usually left to the submitter's appreciation and is usually dependent on the level of "commonality" between the dependent device DTS, as you will see in these examples. The device DTS however should be named consistently with the device model.

kempniu commented:
> If there are enough similarities, create a DTSI with the common bits and include it from the two separate device DTS that will adjust the differences. There are existing examples for Mikrotik (and other) ath79 devices in the repo.
> 
> There is no strong policy for DTSI naming, it's usually left to the submitter's appreciation and is usually dependent on the level of "commonality" between the dependent device DTS, as you will see in these examples. The device DTS however should be named consistently with the device model.

Okay, so it seems reasonable to create a `ar9344_mikrotik_routerboard-95x.dtsi` DTSI with most stuff from this PR in it and then include it in DTS files.

What I am wondering about is *when* to introduce that DTSI: in this PR or in a future PR for the RB951G-2HnD?  If I do it right now, it will look kind of weird to have a DTSI file only included from a single DTS file, so I think it would make more sense from Git history point of view to keep this PR as it is right now and only take care of DTSI reshuffling when I add another similar device, right?

There are more differences between RB951Ui-2HnD and RB951G-2HnD than just switch configuration (e.g. RB951G-2HnD does not support PoE-out and therefore does not have the corresponding LED) and I don't know when I will have enough spare time to identify all of them properly - might be a few days, might be a month or more - and I wouldn't want that uncertainty to delay merging this PR if that is the only outstanding concern about it.

f00b4r0 commented:
> What I am wondering about is _when_ to introduce that DTSI

Don't agonize over this. Do it as you see fit, no one is going to complain either way :)

kempniu commented:
Thanks.  In that case, I acknowledge all the suggestions above and I would propose to leave this PR in its current shape.  I will extract the common parts for RB951Ui-2HnD and RB951G-2HnD to a DTSI file when I have the time to properly identify all the differences between these two devices.

If there is anything else that blocks merging this particular PR, please let me know.

kempniu committed:
ath79: add support for MikroTik RB951Ui-2HnD

MikroTik RB951Ui-2HnD is a wireless SOHO router that was previously
supported by the ar71xx target, see commit d19b868b12 ("ar71xx: Add
support for MikroTik RB951Ui-2HnD").

Specifications
--------------

  - SoC: Atheros AR9344 (600 MHz)
  - RAM: 128 MB (2x 64 MB)
  - Storage: 128 MB NAND flash (various manufacturers)
  - Ethernet: Atheros AR8229 switch, 5x 10/100 Mbit/s
      - 1x PoE in (port 1, 8-30 V input)
      - 1x PoE out (port 5, 500 mA output)
  - Wireless: Atheros AR9340 (802.11b/g/n)
  - USB: 2.0 (1A)
  - 9x LED:
      - 1x power (green, not configurable)
      - 1x user (green)
      - 5x FE ports (green)
      - 1x wireless (green)
      - 1x PoE out (red)
  - 1x button (restart)

See https://mikrotik.com/product/RB951Ui-2HnD for more details.

Flashing
--------

TFTP boot initramfs image and then perform sysupgrade.  Follow
common MikroTik procedures at https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Michał Kępień <openwrt@kempniu.pl>

kempniu commented:
With #12225 now merged, I rebased this PR on top of current `master` and rechecked all the components that this PR claims to support.  Everything still seems to be working as expected.

f00b4r0 commented:
@xback would you consider merging this? :)

mbargo23 commented:
I have compiled the v22.03.4 tag from git and successfully booted the OpenWrt on my RB951Ui-2HnD.
```
[    0.000000] Linux version 5.10.176 (vagrant@localhost.localdomain) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 11.2.0 r19160+962-7ea2f3d6e2) 11.2.0, GNU ld (GNU Binutils) 2.37) #0 Sat Apr 8 16:10:34 2023
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001974c (MIPS 74Kc)
[    0.000000] MIPS: machine is Mikrotik RouterBOARD 951Ui-2HnD
[    0.000000] SoC: Atheros AR9344 rev 2
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000]   Normal zone: 288 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 110792K/131072K available (6182K kernel code, 589K rwdata, 1312K rodata, 10660K init, 205K bss, 20280K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 600.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6370868154 ns
[    0.000010] sched_clock: 32 bits at 300MHz, resolution 3ns, wraps every 7158278654ns
[    0.008921] Calibrating delay loop... 299.82 BogoMIPS (lpj=1499136)
[    0.096062] pid_max: default: 32768 minimum: 301
[    0.101478] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.109779] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.120852] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.132536] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.143748] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.151584] pinctrl core: initialized pinctrl subsystem
[    0.159145] NET: Registered protocol family 16
[    0.176492] gpio-14 (nand_power): hogged as output/high
[    0.236534] clocksource: Switched to clocksource MIPS
[    0.243551] NET: Registered protocol family 2
[    0.248808] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.257757] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.267375] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.276079] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.284114] TCP: Hash tables configured (established 1024 bind 1024)
[    0.291479] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.298966] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.307207] NET: Registered protocol family 1
[    0.312184] PCI: CLS 0 bytes, default 32
[    0.502468] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.514172] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.520864] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.536447] pinctrl-single 1804002c.pinmux: 544 pins, size 68
[    0.543685] gpio-export gpio-export: 2 gpio(s) exported
[    0.550731] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.558745] printk: console [ttyS0] disabled
[    0.563681] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 9, base_baud = 1562500) is a 16550A
[    0.573539] printk: console [ttyS0] enabled
[    0.582610] printk: bootconsole [early0] disabled
[    0.610493] nand: device found, Manufacturer ID: 0xec, Chip ID: 0xf1
[    0.617127] nand: Samsung NAND 128MiB 3,3V 8-bit
[    0.621902] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    0.629779] Scanning device for bad blocks
[    0.640108] Bad eraseblock 49 at 0x000000620000
[    0.700274] Bad eraseblock 512 at 0x000004000000
[    0.757098] Bad eraseblock 946 at 0x000007640000
[    0.771167] 3 fixed-partitions partitions found on MTD device ar934x-nand
[    0.778203] Creating 3 MTD partitions on "ar934x-nand":
[    0.783611] 0x000000000000-0x000000040000 : "booter"
[    0.789859] 0x000000040000-0x000000400000 : "kernel"
[    0.797809] 0x000000400000-0x000008000000 : "ubi"
[    0.811085] spi-nor spi0.0: mx25l512e (64 Kbytes)
[    0.816274] 1 routerbootpart partitions found on MTD device spi0.0
[    0.822710] Creating 1 MTD partitions on "spi0.0":
[    0.827707] 0x000000000000-0x000000010000 : "partitions"
[    0.835484] 4 routerbootpart partitions found on MTD device partitions
[    0.842290] Creating 4 MTD partitions on "partitions":
[    0.847636] 0x000000000000-0x00000000c000 : "routerboot"
[    0.855092] 0x00000000c000-0x00000000d000 : "hard_config"
[    0.861813] 0x00000000d000-0x00000000e000 : "bios"
[    0.869613] 0x00000000e000-0x00000000f000 : "soft_config"
[    0.883310] ag71xx 19000000.eth: invalid MAC address, using random address
[    1.227642] ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
[    1.235786] ag71xx 1a000000.eth: invalid MAC address, using random address
[    1.982491] switch0: Atheros AR8229 rev. 1 switch registered on mdio.0
[    2.064961] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    2.074932] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    2.092647] MikroTik RouterBOARD hardware configuration sysfs driver v0.07
[    2.102248] MikroTik RouterBOARD software configuration sysfs driver v0.05
[    2.110314] NET: Registered protocol family 10
[    2.117216] Segment Routing with IPv6
[    2.121135] NET: Registered protocol family 17
[    2.126078] 8021q: 802.1Q VLAN Support v1.8
[    2.135329] ag71xx 19000000.eth: invalid MAC address, using random address
[    2.478677] ag71xx 19000000.eth: connected to PHY at mdio.0:1f:04 [uid=004dd042, driver=Generic PHY]
[    2.489051] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    2.498135] UBI error: no valid UBI magic found inside mtd2
[    2.553921] Freeing unused kernel memory: 10660K
[    2.558732] This architecture does not have kernel memory protection.
[    2.565398] Run /init as init process
[    2.569189]   with arguments:
[    2.569195]     /init
[    2.569199]   with environment:
[    2.569204]     HOME=/
[    2.569209]     TERM=linux
[    3.093026] init: Console is alive
[    3.097139] init: - watchdog -
[    3.128577] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.154449] usbcore: registered new interface driver usbfs
[    3.160272] usbcore: registered new interface driver hub
[    3.165846] usbcore: registered new device driver usb
[    3.185521] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    3.205936] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.223965] init: - preinit -
[    3.610275] random: jshn: uninitialized urandom read (4 bytes read)
[    3.716701] random: jshn: uninitialized urandom read (4 bytes read)
[    3.972091] random: jshn: uninitialized urandom read (4 bytes read)
[    4.485112] eth0: link up (1000Mbps/Full duplex)
[    4.497350] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    4.504286] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.1: link becomes ready
[    8.694767] eth0: link down
[    8.720510] procd: - early -
[    8.723867] procd: - watchdog -
[    9.336446] procd: - watchdog -
[    9.341378] procd: - ubus -
[    9.354458] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.394852] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.402054] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.413086] procd: - init -
[   10.240165] kmodloader: loading kernel modules from /etc/modules.d/*
[   10.662601] urngd: v1.0.2 started.
[   10.827569] Loading modules backported from Linux version v5.15.92-0-ge515b9902f5f
[   10.835404] Backport generated by backports.git v5.15.92-1-0-gdfe0f60c
[   10.882944] random: crng init done
[   10.886479] random: 26 urandom warning(s) missed due to ratelimiting
[   11.029976] PPP generic driver version 2.4.2
[   11.038381] NET: Registered protocol family 24
[   11.095306] ath9k 18100000.wmac: Direct firmware load for ath9k-eeprom-ahb-18100000.wmac.bin failed with error -2
[   11.105978] ath9k 18100000.wmac: Falling back to sysfs fallback for: ath9k-eeprom-ahb-18100000.wmac.bin
[   11.275672] ath: EEPROM regdomain: 0x0
[   11.275686] ath: EEPROM indicates default country code should be used
[   11.275691] ath: doing EEPROM country->regdmn map search
[   11.275708] ath: country maps to regdmn code: 0x3a
[   11.275716] ath: Country alpha2 being used: US
[   11.275722] ath: Regpair used: 0x3a
[   11.289157] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   11.291339] ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=12
[   11.337069] kmodloader: done loading kernel modules from /etc/modules.d/*
[   48.111530] eth0: link up (1000Mbps/Full duplex)
[   48.116336] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   48.143131] device eth0 entered promiscuous mode
[   48.149387] br-lan: port 1(eth0.1) entered blocking state
[   48.154983] br-lan: port 1(eth0.1) entered disabled state
[   48.160940] device eth0.1 entered promiscuous mode
[   48.193779] br-lan: port 1(eth0.1) entered blocking state
[   48.199422] br-lan: port 1(eth0.1) entered forwarding state
[   49.147042] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   50.429578] eth1: link up (100Mbps/Full duplex)
[   50.434313] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready

Please press Enter to activate this console.

BusyBox v1.35.0 (2023-04-08 16:10:34 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt 22.03-SNAPSHOT, r19160+962-7ea2f3d6e2
 -----------------------------------------------------
=== WARNING! =====================================
There is no root password defined on this device!
Use the "passwd" command to set up a new password
in order to prevent unauthorized SSH logins.
--------------------------------------------------
root@OpenWrt:/#
```

There is a problem with fetching the MAC address of `eth0`. On the sticker attached to the back of my unit there are two MACs: `D4:CA:6D:DC:E0:FA` and `D4:CA:6D:DC:E0:FF`. Interface `eth1` does have the correct address. Any tips what to check or change to get the correct MAC of `eth0`?

```
root@OpenWrt:/# ip link show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether 76:24:f2:81:d8:71 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP qlen 1000
    link/ether d4:ca:6d:dc:e0:fa brd ff:ff:ff:ff:ff:ff
5: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether d4:ca:6d:dc:e1:05 brd ff:ff:ff:ff:ff:ff
6: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether d4:ca:6d:dc:e0:fb brd ff:ff:ff:ff:ff:ff
7: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether d4:ca:6d:dc:e0:fb brd ff:ff:ff:ff:ff:ff
```

neheb commented:
@kempniu any reason why qca,led-pin is used?

kempniu commented:
@neheb: I assume you ask because `target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts` is the only user of that property in the OpenWRT tree - I was not aware of that until you pointed it out :-)

The reason I used `qca,led-pin` is because:

  - by default, the ath9k driver [assigns](https://elixir.bootlin.com/linux/v6.9.9/source/drivers/net/wireless/ath/ath9k/gpio.c#L40) the [default GPIO number](https://elixir.bootlin.com/linux/v6.9.9/source/drivers/net/wireless/ath/ath9k/ath9k.h#L838) (1) to the WLAN LED and that's not the number that the 951Ui uses (11), so it doesn't work out-of-the-box (FTR, none of the custom OpenWRT patches for the ath9k driver fix that, either),
  - since the ath9k driver registers a LED anyway (`/sys/class/leds/ath9k-phy0`), I was trying to avoid having two LEDs for WLAN activity, one of which does not work...

I confirmed that a patch like this one also makes the WLAN LED work on the 951Ui, but it still leaves the duplicate `ath9k-phy0` LED with the wrong GPIO number assigned in sysfs...

```patch
diff --git a/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts b/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts
index 65003cb9a1..e61eed8ae5 100644
--- a/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts
+++ b/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts
@@ -40,6 +40,12 @@
 			label = "green:port5";
 			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
 		};
+
+		wlan {
+			label = "green:wlan";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
 	};
 
 	gpio-export {
@@ -68,7 +74,3 @@
 &eth1 {
 	compatible = "qca,ar9340-eth", "syscon";
 };
-
-&wmac {
-	qca,led-pin = /bits/ 8 <11>;
-};
```

f00b4r0 commented:
I think there might be merit in using the `qca,led-pin` trick on all these devices (same for ath10k if it's available btw), unless there is a different way to disable registration of the athN-wlan LEDs. The non-working extraneous athNk-wlan LEDs have always bothered me :)
My 2c

neheb commented:
Well that’s an unfortunate answer. And here I am trying to get rid of it.

I would think an led-pin value of -1 disables it.

john-tho commented:
> I confirmed that a patch like this one also makes the WLAN LED work on the 951Ui, but it still leaves the duplicate `ath9k-phy0` LED with the wrong GPIO number assigned in sysfs...

I have not tried it, but I saw this mentioned some weeks ago looking into multiple leds on ath10k gpios: https://lore.kernel.org/linux-wireless/4345c367-e786-ef44-fed7-8658f519a689@kresin.me/ (the v12 ath10k LEDS (before GPIOs were dropped from it)).
https://github.com/openwrt/openwrt/blob/main/package/kernel/mac80211/patches/ath9k/548-ath9k_enable_gpio_chip.patch
